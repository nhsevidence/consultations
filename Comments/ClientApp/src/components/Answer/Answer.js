import React, { Component, Fragment } from "react";
import Moment from "react-moment";

import stringifyObject from "stringify-object";

type StatusType = {
	statusId: number,
	name: string
};

type AnswerType = {
	answerId: number,
	answerText: string,
	answerBoolean: boolean,
	questionId: number,
	lastModifiedDate: Date,
	lastModifiedByUserId: string,
	statusId: number,
	status: StatusType
};

type PropsType = {
	staticContext?: any,
	isVisible: boolean,
	answer: AnswerType,
	readOnly: boolean,
	saveHandler: Function,
	deleteHandler: Function,
	unique: string
};

type StateType = {
	answerId: number,
	answerText: string,
	answer: AnswerType
};

export class Answer extends Component<PropsType, StateType> {
	constructor() {
		super();
		this.state = {
			answer: {
				answerText: ""
			},
			unsavedChanges: false
		};
	}

	componentDidMount() {
		this.setState({
			answer: this.props.answer
		});
	}

	textareaChangeHandler = e => {
		const answer = this.state.answer;
		answer.answerText = e.target.value;
		this.setState({
			answer,
			unsavedChanges: true
		});
	};

	static getDerivedStateFromProps(nextProps, prevState) {
		 console.log(`answer nextProps: ${stringifyObject(nextProps)}`);
		 console.log(`answer prevState: ${stringifyObject(prevState)}`);
		console.log('get derived thing in answer');
		const prevTimestamp = prevState.answer.lastModifiedDate;
		const nextTimestamp = nextProps.answer.lastModifiedDate;
		const hasAnswerBeenUpdated = () => prevTimestamp !== nextTimestamp;
		if (hasAnswerBeenUpdated()) {
			return {
				answer: nextProps.answer,
				unsavedChanges: false
			};
		}
		return null;
	}

	render() {
		if (!this.state.answer) return null;
		const {
			answerText,
			lastModifiedDate,
			answerId,
			questionId
		} = this.state.answer;
		const unsavedChanges = this.state.unsavedChanges;
		const answer = this.state.answer;
		const readOnly = this.props.readOnly;
		const moment = require("moment");

		return (

			<Fragment>
				<section role="form">

					{lastModifiedDate ? (
						<div className="CommentBox__datestamp mb--d font-weight-bold">
							Last Modified:{" "}
							<Moment format="D/M/YYYY - h:mma" date={moment.utc(lastModifiedDate).toDate()}/>
						</div>
					) : null}
					<form onSubmit={e => this.props.saveHandler(e, answer)}>
						<div className="form__group form__group--textarea mb--0">
							<textarea
								data-qa-sel="Comment-text-area"
								disabled={readOnly}
								id={this.props.unique}
								className="form__input form__input--textarea"
								onChange={this.textareaChangeHandler}
								placeholder="Enter your answer here"
								value={answerText}/>
						</div>
						{!readOnly && answerText && answerText.length > 0 && (
							<input
								data-qa-sel="submit-button"
								className="btn ml--0"
								type="submit"
								value={unsavedChanges ? "Save answer" : "Saved"}
								disabled={!unsavedChanges}/>

						)}
						{!readOnly &&
						<button
							data-qa-sel="delete-comment-button"
							className="btn mr--0 right"
							onClick={e => this.props.deleteHandler(e, questionId, answerId)}>
							<span className="visually-hidden">Delete this answer</span>
							<span className="icon icon--trash" aria-hidden="true"/>
						</button>
						}
					</form>
				</section>
			</Fragment>
		);
	}
}

export default Answer;
