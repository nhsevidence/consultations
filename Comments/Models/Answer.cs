using System;
using System.Collections.Generic;

namespace Comments.Models
{
    public partial class Answer
    {
        private Answer() { } //just for EF

        public Answer(int questionId, Guid createdByUserId, string answerText, bool? answerBoolean, Question question)
        {
            QuestionId = questionId;
            CreatedByUserId = createdByUserId;
            AnswerText = answerText ?? throw new ArgumentNullException(nameof(answerText));
            AnswerBoolean = answerBoolean;
            Question = question;
        }

        public void UpdateFromViewModel(ViewModels.Answer answer)
        {
            LastModifiedByUserId = answer.LastModifiedByUserId;
            AnswerText = answer.AnswerText ?? throw new ArgumentNullException(nameof(answer.AnswerText)); 
            AnswerBoolean = answer.AnswerBoolean;
        }
    }
}
