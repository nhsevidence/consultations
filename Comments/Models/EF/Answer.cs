using System;
using System.Collections.Generic;

namespace Comments.Models
{
    public partial class Answer
    {
        public int AnswerId { get; set; }
		public int QuestionId { get; set; }
		public string CreatedByUserId { get; set; }
		public string AnswerText { get; set; }
		public bool? AnswerBoolean { get; set; }
        public DateTime CreatedDate { get; set; }
		public DateTime LastModifiedDate { get; set; }
		public string LastModifiedByUserId { get; set; }
		public int StatusId { get; set; }
        public int? OrganisationUserId { get; set; }
        public int? OrganisationAuthorisationId { get; set; }
        public int? ParentAnswerId { get; set; }

        public OrganisationAuthorisation OrganisationAuthorisation { get; set; }
        public OrganisationUser OrganisationUser { get; set; }
        public Answer ParentAnswer { get; set; }
        public Question Question { get; set; }
        public Status Status { get; set; }
        public ICollection<Answer> ChildAnswers { get; set; }
        public ICollection<SubmissionAnswer> SubmissionAnswer { get; set; }
    }
}
