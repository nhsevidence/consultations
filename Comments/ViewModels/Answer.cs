using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Comments.ViewModels
{
	public class Answer
	{
		public Answer() { } //only here for model binding. don't use it in code.
		public Answer(int answerId, string answerText, bool answerBoolean, DateTime lastModifiedDate, Guid lastModifiedByUserId, int questionId, int statusId)
		{
			AnswerId = answerId;
			AnswerText = answerText;
			AnswerBoolean = answerBoolean;
			QuestionId = questionId;
			LastModifiedDate = lastModifiedDate;
			LastModifiedByUserId = lastModifiedByUserId;
			StatusId = statusId;
		}

		public Answer(Models.Answer answer)
		{
			AnswerId = answer.AnswerId;
			AnswerText = answer.AnswerText;
			AnswerBoolean = answer.AnswerBoolean;
			QuestionId = answer.QuestionId;
			LastModifiedDate = answer.LastModifiedDate;
			LastModifiedByUserId = answer.LastModifiedByUserId;
			StatusId = 1; //answer.StatusId;
			Status = null; //new Status(answer.Status);
		}

		public int AnswerId { get; set; }
		public string AnswerText { get; set; }
		public bool? AnswerBoolean { get; set; }

		public int QuestionId { get; set; }
		public DateTime LastModifiedDate { get; set; }
		public Guid LastModifiedByUserId { get; set; }

		public int StatusId { get; set; }
		public ViewModels.Status Status { get; set; }
	}
}
