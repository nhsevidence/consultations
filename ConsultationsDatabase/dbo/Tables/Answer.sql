CREATE TABLE [dbo].[Answer] (
    [AnswerID]         INT              IDENTITY (1, 1) NOT NULL,
    [QuestionID]       INT              NOT NULL,
    [CreatedByUserID]           UNIQUEIDENTIFIER NOT NULL,
    [AnswerText]       NVARCHAR (MAX)   NULL,
    [AnswerBoolean]    BIT              NULL,
    [CreatedDate]      DATETIME2 (7)    CONSTRAINT [DF_Answer_CreatedDate] DEFAULT (getdate()) NOT NULL,
    [LastModifiedDate] DATETIME2 (7)    CONSTRAINT [DF_Answer_LastModifiedDate] DEFAULT (getdate()) NOT NULL,
    [LastModifiedByUserID] UNIQUEIDENTIFIER NOT NULL, 
    [IsDeleted] BIT NOT NULL DEFAULT 0, 
    CONSTRAINT [PK_Answer] PRIMARY KEY CLUSTERED ([AnswerID] ASC),
    CONSTRAINT [FK_Answer_Question] FOREIGN KEY ([QuestionID]) REFERENCES [dbo].[Question] ([QuestionID])
);

