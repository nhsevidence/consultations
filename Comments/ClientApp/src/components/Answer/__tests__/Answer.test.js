/* global jest */
import React from "react";
import { mount } from "enzyme";
import { Answer } from "../Answer";
import answerWithAnswer from "./answerWithAnswer.json";
import answerWithoutAnswer from "./answerWithoutAnswer.json";
import toJson from "enzyme-to-json";

describe("[ClientApp] ", () => {
	describe("Answer Component", () => {
		const answerPropsWithAnswer = {
			isVisible: true,
			answer: answerWithAnswer,
			readOnly: false,
			saveHandler: jest.fn(),
			deleteHandler: jest.fn(),
			unique: "string",
			updateUnsavedIds: jest.fn(),
			questionType: {
				type: "Text",
			},
		};

		const answerPropsWithoutAnswer = {
			isVisible: true,
			answer: answerWithoutAnswer,
			readOnly: false,
			saveHandler: jest.fn(),
			deleteHandler: jest.fn(),
			unique: "string",
			updateUnsavedIds: jest.fn(),
			questionType: {
				type: "Text",
			},
		};

		it("sets text area with comment text correctly", async () => {
			const wrapper = mount(<Answer {...answerPropsWithAnswer} />);
			expect(wrapper.find("textarea").length).toEqual(1);
			expect(wrapper.find("textarea").props().defaultValue).toEqual(
				"some answer text",
			);
		});

		it("unsavedChanges function is fired correctly on text area change", () => {
			const wrapper = mount(<Answer {...answerPropsWithAnswer} />);
			expect(wrapper.state().unsavedChanges).toEqual(false);
			const textArea = wrapper.find("textarea");
			textArea.simulate("input", {
				target: {
					value: "an updated answer",
				},
			});
			expect(wrapper.state().answer.answerText).toEqual("an updated answer");
			expect(answerPropsWithAnswer.updateUnsavedIds).toHaveBeenCalledWith(
				"22q",
				true,
			);
		});

		it("should update UnsavedChanges if lastupdateddate has changed", () => {
			const wrapper = mount(<Answer {...answerPropsWithAnswer} />);
			wrapper.setState({ unsavedChanges: true });
			const updatedProps = {
				answer: {
					answerId: answerWithoutAnswer.answerId,
					answerText: "an updated answer",
					lastModifiedDate: new Date("02/04/2018").toISOString(),
				},
			};
			wrapper.setProps(updatedProps);
			expect(wrapper.state().unsavedChanges).toEqual(false);
		});

		it("should not update UnsavedChanges if lastupdateddate has not changed", () => {
			const wrapper = mount(<Answer {...answerPropsWithAnswer} />);
			wrapper.setState({ unsavedChanges: true });
			wrapper.setProps(answerPropsWithAnswer);
			expect(wrapper.state().unsavedChanges).toEqual(true);
		});

		it("updated comment text in state after new props received", () => {
			const wrapper = mount(<Answer {...answerPropsWithAnswer} />);
			const updatedProps = {
				answer: {
					answerId: answerWithoutAnswer.answerId,
					answerText: "an updated answer",
				},
			};
			wrapper.setProps(updatedProps);
			expect(wrapper.state().answer.answerText).toEqual("an updated answer");
		});

		it("should match snapshot with answer", () => {
			const wrapper = mount(<Answer {...answerPropsWithAnswer} />);
			expect(
				toJson(wrapper, {
					noKey: true,
					mode: "deep",
				}),
			).toMatchSnapshot();
		});

		it("should match snapshot without answer", () => {
			const wrapper = mount(<Answer {...answerPropsWithoutAnswer} />);
			expect(
				toJson(wrapper, {
					noKey: true,
					mode: "deep",
				}),
			).toMatchSnapshot();
		});
	});
});
