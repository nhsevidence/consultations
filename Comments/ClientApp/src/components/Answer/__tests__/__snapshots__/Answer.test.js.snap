// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ClientApp]  Answer Component should match snapshot with answer 1`] = `
<section
  role="form"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form__group form__group--textarea mb--0"
    >
      <textarea
        className="form__input form__input--textarea"
        data-qa-sel="Comment-text-area"
        disabled={false}
        id="string"
        onChange={[Function]}
        placeholder="Enter your answer here"
        value="an updated answer"
      />
    </div>
    <span
      className="ml--0 CommentBox__savedIndicator"
    >
      Saved
    </span>
    <button
      className="btn mr--0 right"
      data-qa-sel="delete-comment-button"
      onClick={[Function]}
    >
      <span
        className="visually-hidden"
      >
        Delete this answer
      </span>
      <span
        aria-hidden="true"
        className="icon icon--trash"
      />
    </button>
  </form>
</section>
`;

exports[`[ClientApp]  Answer Component should match snapshot without answer 1`] = `
<section
  role="form"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form__group form__group--textarea mb--0"
    >
      <textarea
        className="form__input form__input--textarea"
        data-qa-sel="Comment-text-area"
        disabled={false}
        id="string"
        onChange={[Function]}
        placeholder="Enter your answer here"
        value=""
      />
    </div>
  </form>
</section>
`;
